### PR TITLE
CakePHP ^4.2 and RR 2.7

### PR DESCRIPTION
**Pull request by cnizzardini**
This PR makes this plugin compatible with CakePHP 4.3 and RR 2.7.

Still needs:
- Unit Tests
- More testing
- Maybe needs version 0.2 and composer to list not compatible with 0.1

### Automated Testing

Run test suite + static analysis, see composer.json `scripts`.

```console
composer analyze
```

### Manual Testing:

```console
composer create-project --prefer-dist cakephp/app
```
Add vcs source to composer:

```json
    "repositories": [
        {
            "type": "vcs",
            "url":  "git@github.com:cnizzardini/cakephp-roadrunner.git"
        }
    ]
```

Then from context of your project root directory:

```console
composer require cakedc/cakephp-roadrunner:dev-cakephp42-roadrunner-27
cp vendor/cakedc/cakephp-roadrunner/worker/cakephp-worker.php .
cp vendor/cakedc/cakephp-roadrunner/worker/rr.yaml .
```

Download RR binary and then run server, example:

```console
/usr/local/bin/rr serve -d -c rr.yaml
```

Browse to http://localhost:8080

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,38 +1,81 @@
...
+## Install
 
-```bash
+Install via composer:
+
+```console
+composer require cakedc/cakephp-roadrunner
+```
+
+Unlike most CakePHP plugins you won't be needing to load the plugin in your `src/Application.php`.
```
Comment (ADmad) [inline README.md:31]: Probably doesn't matter much but if this doesn't need to be loaded as a plugin then having type `cakephp-plugin` in composer.json is not necessary.

File: composer.json [hunk 2]
```diff
@@ -4,12 +4,13 @@
     "type": "cakephp-plugin",
     "license": "MIT",
     "require": {
-        "cakephp/cakephp": "^3.6",
-        "spiral/roadrunner": "^1.3",
-        "relay/middleware": "^1.2"
+        "php": "^8.0",
+        "cakephp/cakephp": "^4.3",
+        "spiral/roadrunner": "^2.7",
+        "nyholm/psr7": "^1.4"
```
Comment (ADmad) [inline composer.json]: Why is this package needed? Cake already has PSR-7 complaint classes.

Comment (cnizzardini) [reply composer.json]: Good catch, its from the roadrunner examples.

Comment (ADmad) [reply composer.json]: You'll probably be able to get rid of the `convertRequest()` method too from `Bridge` if you pass Cake's `ServerRequest` from the worker file.

File: README.md (hunk 1 above)
Comment (cnizzardini) [reply README.md:31]: I'd say just leave it.

File: src/Http/ServerRequestFactory.php [hunk 3]
```diff
@@ -2,22 +2,33 @@
...
-use Cake\Error\Debugger;
 use Cake\Http\ServerRequest;
+use Cake\Http\ServerRequestFactory as BaseServerRequestFactory;
+use function Laminas\Diactoros\normalizeServer;
 
-class ServerRequestFactory extends \Cake\Http\ServerRequestFactory
+/**
+ * Creates an instance of ServerRequest.
+ */
+class ServerRequestFactory extends BaseServerRequestFactory
```
Comment (ADmad) [inline src/Http/ServerRequestFactory.php:14]: This class is currently only used in tests, `Bridge` directly uses `Cake\Http\ServerRequestFactory`. So is it really required?

File: composer.json (hunk 2 above)
Comment (cnizzardini) [reply composer.json]: I'm not seeing Cake implementations of the following:

- Psr\Http\Message\StreamFactoryInterface
- Psr\Http\Message\UploadedFileFactoryInterface

Seems easier just to use that package and then also potentially delete ServerRequestFactory

File: src/Http/ServerRequestFactory.php (hunk 3 above)
Comment (cnizzardini) [reply src/Http/ServerRequestFactory.php:14]: I don't think so, it does make use of `CakeDC\Roadrunner\Http\Session` though. I haven't gotten to testing sessions and an actual real cakephp app with this though, just testing on cakephp welcome page so far.

File: composer.json (hunk 2 above)
Comment (ADmad) [reply composer.json]: We use laminas/diactoros' classes for those interfaces.

Is explicitly creating instances for those even required? It's handled internally in the `ServerRequest` class.

File: src/Bridge.php [hunk 4]
```diff
@@ -1,117 +1,109 @@
...
+    public function handle(ServerRequestInterface $request): ResponseInterface
     {
-        $response = new \Cake\Http\Response();
-        $request = $this->convertRequest($request);
-        $middleware = $this->application->middleware(new \Cake\Http\MiddlewareQueue());
-        if ($this->application instanceof \Cake\Core\PluginApplicationInterface) {
-            $middleware = $this->application->pluginMiddleware($middleware);
-        }
+        $middleware = $this->application->middleware(new MiddlewareQueue());
+        $middleware = $this->application->pluginMiddleware($middleware);
```
Comment (cnizzardini) [inline src/Bridge.php]: Not sure what purpose this serves @ADmad @steinkel . Also phpstan complains:

```
> phpstan analyse src/
Note: Using configuration file /var/www/personal/cakephp-roadrunner/phpstan.neon.
 4/4 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------------------------------------- 
  Line   Bridge.php                                                                           
 ------ ------------------------------------------------------------------------------------- 
  87     Call to an undefined method Cake\Core\HttpApplicationInterface::pluginMiddleware().  
 ------ ------------------------------------------------------------------------------------- 
```

Comment (cnizzardini) [reply src/Bridge.php]: Sidenote: I wonder if a plugin should be added to test_app to verify plugin middleware is ran...

File: composer.json (hunk 2 above)
Comment (cnizzardini) [reply composer.json]: Thanks, thats resolved now.

Comment (cnizzardini) [reply composer.json]: Hmm... except it no longer works when those are replaced!

Edit: `convertRequest()` is doing something important.

File: src/Bridge.php (hunk 4 above)
Comment (ADmad) [reply src/Bridge.php]: The phpstan error is because `pluginMiddleware()` method is part of `PluginApplicationInterface` where as as `Bridge::$application` is typed as just `HttpApplicationInterface`. So you need to wrap the call in a `instanceof PluginApplicationInterface` check. 

This `handle()` seems to do pretty much the same thing that [Server::run()](https://github.com/cakephp/cakephp/blob/41ce80827fd964f6b32e799c837fc3c585a097c1/src/Http/Server.php#L75) does. So either try just calling that or copy it and modify as required.

File: composer.json (hunk 2 above)
Comment (ADmad) [reply composer.json]: Maybe the environment vars it sets (using `$server`) need to be set explicitly. Try to narrow down what's the relevant stuff it does and then directly modify the `$request` instance. Recreating the request instance again surely shouldn't be necessary if it's already a `Cake\Http\ServerRequest` instance.

**Comment by cnizzardini** [comment]
Another cute problem I've been looking at, `BodyParserMiddleware` is not handling json request body on something basic like this: 

```console
curl -X 'POST' \
  'http://localhost:8080/actors' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "first_name": "string",
  "last_name": "string"
}'
```

The stream is empty at this point: https://github.com/cakephp/cakephp/blob/4.x/src/Http/Middleware/BodyParserMiddleware.php#L166

**Edit:** Resolved by forcing the request body into the cakephp request for now, added unit tests for post, put patch.
```php
        $body = $request->getBody();
        $cakeRequest = clone $cakeRequest->withBody($body);
```

File: src/Bridge.php (hunk 4 above)
Comment (cnizzardini) [reply src/Bridge.php]: As best I can tell, cake internals and various middleware require `Cake\Http\ServerRequest` in a number of places. For instance things like this fail:

```
Error Object
(
    [message:protected] => Call to undefined method Laminas\Diactoros\ServerRequest::clientIp()
    [string:Error:private] => 
    [code:protected] => 0
    [file:protected] => /var/www/personal/cakephp-swagger-bake-demo/vendor/dereuromark/cakephp-setup/src/Middleware/MaintenanceMiddleware.php
    [line:protected] => 48
```

```
{"title":"CakeDC Roadrunner Error","exception":"TypeError","message":"Cake\\Routing\\Router::setRequest(): Argument #1 ($request) must be of type Cake\\Http\\ServerRequest, Laminas\\Diactoros\\ServerRequest given, called in \/var\/www\/personal\/cakephp-swagger-bake-demo\/vendor\/cakephp\/cakephp\/src\/Routing\/Middleware\/RoutingMiddleware.php on line 135","trace":null}
```

I will just clean up some of the cake that converts to cake server request for now.

File: composer.json (hunk 2 above)
Comment (cnizzardini) [reply composer.json]: As best I can tell, cake internals and various middleware require `Cake\Http\ServerRequest` in a number of places. For instance things like this fail:

```
Error Object
(
    [message:protected] => Call to undefined method Laminas\Diactoros\ServerRequest::clientIp()
    [string:Error:private] => 
    [code:protected] => 0
    [file:protected] => /var/www/personal/cakephp-swagger-bake-demo/vendor/dereuromark/cakephp-setup/src/Middleware/MaintenanceMiddleware.php
    [line:protected] => 48
```

```
{"title":"CakeDC Roadrunner Error","exception":"TypeError","message":"Cake\\Routing\\Router::setRequest(): Argument #1 ($request) must be of type Cake\\Http\\ServerRequest, Laminas\\Diactoros\\ServerRequest given, called in \/var\/www\/personal\/cakephp-swagger-bake-demo\/vendor\/cakephp\/cakephp\/src\/Routing\/Middleware\/RoutingMiddleware.php on line 135","trace":null}
```

I only encountered these errors after changing ExceptionRenderer to accept Laminos instead of Cake ServerRequest.

I will just clean up some of the code that converts to cake server request for now.

Comment (ADmad) [reply composer.json]: Using `Cake\Http\ServerRequestFactory` instead of `Laminas\Diactoros\ServerRequestFactory` in the worker file would avoid this problem and remove the need for `Bridge::convertRequest()` too.

But for some reason `Cake\Http\ServerRequestFactory` is made `abstract` so it can't be instantiated. No idea why that is. Perhaps something we should change in the core.

Comment (ADmad) [reply composer.json]: So this is why `CakeDC\Roadrunner\Http\ServerRequestFactory` was required and you should use that in the worker file :slightly_smiling_face:

File: src/Bridge.php [hunk 5]
```diff
@@ -1,117 +1,137 @@
...
-
-        $cakeRequest = ServerRequestFactory::fromGlobals($server, $query, $body, $cookies, $files);
+        $cakeRequest = ServerRequestFactory::fromGlobals(
+            $request->getServerParams(),
+            $request->getQueryParams(),
+            $request->getParsedBody(),
+            $request->getCookieParams(),
+            $request->getUploadedFiles()
+        );
         $cakeRequest->trustProxy = true;
```
Comment (cnizzardini) [inline src/Bridge.php:133]: Why is this needed @steinkel ?

Comment (steinkel) [reply src/Bridge.php:133]: The main reason was providing translation between \Zend\Diactoros\ServerRequest into \Cake\Http\ServerRequest, it was not possible to pass the request without conversion from/to the cakephp stack for processing without this utility in those times...

Comment (cnizzardini) [reply src/Bridge.php:133]: Specifically I mean the `$cakeRequest->trustProxy = true;`

Comment (steinkel) [reply src/Bridge.php:133]: I barely remember it was needed to correctly pass the correct IP to cake context, as the go server was acting as a proxy
